File: NinePSharp.Messages.FSharp/messages/Tversion.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tversion =
    { Size: uint32
      Tag: uint16
      MSize: uint32
      Version: string }
    
    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tversion
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tversion span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.MSize); offset <- offset + 4
            writeString this.Version span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let msize = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let version = readString data &offset
        { Size = size; Tag = tag; MSize = msize; Version = version }

File: NinePSharp/messages/Tversion.cs
using System.Buffers.Binary;
using NinePSharp.Interfaces;
using NinePSharp.Constants;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

public readonly struct Tversion : ISerializable
{
    public uint Size {get;}
    public MessageTypes Type => MessageTypes.Tversion;
    public ushort Tag {get;}
    public uint MSize {get;}
    public string Version {get;}

    public Tversion(ushort tag, uint msize, string version)
    {
        Tag = tag;
        MSize = msize;
        Version = version;
        Size = (uint)(NinePConstants.HeaderSize + 4 + 2 + System.Text.Encoding.UTF8.GetByteCount(version));
    }

    public Tversion(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        MSize = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
        offset += 4;
        Version = data.ReadString(ref offset);
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), MSize);
        offset += 4;
        data.WriteString(Version, ref offset);
    }
}